"""Bayesian variant calling with FreeBayes.

http://bioinformatics.bc.edu/marthlab/FreeBayes
"""
import os
import shutil
import subprocess

from bcbio import broad
from bcbio.utils import file_exists
from bcbio.distributed.transaction import file_transaction
from bcbio.variation import annotation, genotype
from bcbio.log import logger

def run_freebayes(align_bam, ref_file, config, dbsnp=None, region=None,
                  out_file=None):
    """Detect small polymorphisms with FreeBayes.
    """
    if out_file is None:
        out_file = "%s-variants.vcf" % os.path.splitext(align_bam)[0]
    if not file_exists(out_file):
        logger.info("Genotyping with FreeBayes: {region} {fname}".format(
            region=region, fname=os.path.basename(align_bam)))
        with file_transaction(out_file) as tx_out_file:
            cl = [config["program"].get("freebayes", "freebayes"),
                  "-b", align_bam, "-v", tx_out_file, "-f", ref_file]
            if region:
                cl.extend(["-r", region])
            try:
                subprocess.check_call(cl)
            # XXX Temporary, work around freebayes issue; need to recall these regions
            # later so this is an ugly silent fix. Will need to grep for 'freebayes failed'
            # https://github.com/ekg/freebayes/issues/22
            except subprocess.CalledProcessError:
                with open(tx_out_file, "w") as out_handle:
                    out_handle.write("##fileformat=VCFv4.1\n"
                                     "## No variants; freebayes failed\n"
                                     "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\n")
    return out_file

def postcall_filter(in_file, ref_file, vrn_files, config):
    """Perform quality score filtering on FreeBayes variant calls.
    """
    out_file = _check_file_gatk_merge(in_file)
    out_file = annotation.annotate_dbsnp(out_file, vrn_files.dbsnp,
                                         ref_file, config)
    filters = ["QUAL < 20.0", "DP < 5"]
    out_file = genotype.variant_filtration_with_exp(broad.runner_from_config(config),
                                                    out_file, ref_file, "", filters)
    return out_file

def _check_file_gatk_merge(vcf_file):
    """Remove problem lines generated by GATK merging from FreeBayes calls.

    Works around this issue until next GATK release:
    http://getsatisfaction.com/gsa/topics/
    variantcontext_creates_empty_allele_from_vcf_input_with_multiple_alleles
    """
    def _not_empty_allele(line):
        parts = line.split("\t")
        alt = parts[4]
        return not alt[0] == ","
    orig_file = "{0}.orig".format(vcf_file)
    if not file_exists(orig_file):
        shutil.move(vcf_file, orig_file)
        with open(orig_file) as in_handle:
            with open(vcf_file, "w") as out_handle:
                for line in in_handle:
                    if line.startswith("#") or _not_empty_allele(line):
                        out_handle.write(line)
    return vcf_file
